MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c

```diff
@@ -271,6 +277,23 @@ EnumerateNvmeDevNamespace (
       goto Exit;
     }
 
+    // MU_CHANGE [BEGIN] - Request Number of Queues from Controller
+    if (NVME_SUPPORT_BLOCKIO2 (Private)) {
+      // We have multiple data queues, so we can support the BlockIo2 protocol
+      Status = gBS->InstallMultipleProtocolInterfaces (
+                      &Device->DeviceHandle,
+                      &gEfiBlockIo2ProtocolGuid,
+                      &Device->BlockIo2,
+                      NULL
+                      );
+      if (EFI_ERROR (Status)) {
+        DEBUG ((DEBUG_ERROR, "%a: Failed to install BlockIo2 protocol. Error %r\n", __func__, Status));
+        goto Exit;

```

Comment:
At the Point that Blockio2 attempts to install, if an error occurs, BlockIo et all are still installed on the handle, but jumping to Exit will leave the name spaces for all the nvme controllers un enumerated.

What is the best way to handle this scenario? Do the BlockIo et all need uninstalled? Should the system downgrade the Number of Io pairs and allow it to limp along just as on BlockIo?

Reply:
Since BlockIo2 is not heavily used, probably trying to continue on (with an assert) makes the most sense?

Reply:
In NvmExpress.c ProcessAsyncTaskList, if any of the subtasks fail it won't process subsequent ones. I'm not sure if this is the desired behavior. Before this change, BlockIo2 protocol was installed with BlockIo and if failed, all protocols would be uninstalled.

Reply:
Issue created: https://github.com/microsoft/mu_basecore/issues/1358